### PR TITLE
Fix schedules

### PR DESCRIPTION
**Pull request by schai6**
Added all api and changed schedules minus attendance api

### CODE REVIEW

File: ap/accounts/models.py [hunk 1]
```diff
@@ -49,189 +53,284 @@
...
+    weeks = range(start_week, end_week + 1)
+    w_tb=OrderedDict()
+    # for every schedule, filter events to get events in the date range.
+    for schedule in schedules:
+      evs = schedule.events.filter(weekday__gte=start.weekday(), weekday__lte=end.weekday())
+      # create week table
+      w_tb = self.compute_prioritized_event_table(w_tb, weeks, evs)
+    # create event list.
+    return self.export_event_list_from_table(w_tb)
+    # TODO: Think more about this
```
Comment (radicalray) [inline ap/accounts/models.py]: Delete

File: ap/accounts/models.py [hunk 2]
```diff
@@ -49,189 +53,284 @@
...
+  def export_event_list_from_table(self, w_tb):
+    event_list=[]
+    for (w, d), evs in w_tb.items():
+      for ev in evs:
+        date = ev.date_for_week(w)
+        # calc date from w
+        ev.start_datetime = datetime.combine(date, ev.start)
+        ev.end_datetime = datetime.combine(date, ev.end)
+        # append a copy of ev to answer list you will return
+        event_list.append(copy(ev))
```
Comment (radicalray) [inline ap/accounts/models.py]: Copy is expensive.

File: ap/ap/settings/base.py [hunk 3]
```diff
@@ -145,6 +145,7 @@
     'paintstore',
     'solo',
     'django_extensions',
+    'rest_framework_swagger',
```
Comment (radicalray) [inline ap/ap/settings/base.py]: Is this necessary?

File: ap/ap/views.py [hunk 4]
```diff
@@ -19,7 +19,7 @@ def home(request):
     if is_trainee(request.user):
         trainee = trainee_from_user(request.user)
         try:
-            data['schedule'] = trainee.schedule.get(term=Term.current_term())
+            data['schedules'] = trainee.schedules.all()
```
Comment (radicalray) [inline ap/ap/views.py]: Schedule for current term instead of all schedules?

File: ap/attendance/views.py [hunk 5]
```diff
@@ -67,19 +56,20 @@ class RollViewSet(BulkModelViewSet):
     def get_queryset(self):
         user = self.request.user
         trainee = trainee_from_user(user)
-        roll = Roll.objects.filter(trainee=(trainee), event__term=Term.current_term())
+        roll = trainee.current_rolls
         return roll
     def allow_bulk_destroy(self, qs, filtered):
-        return not all(x in filtered for x in qs)
+        return filtered
+        # return not all(x in filtered for x in qs)
```
Comment (radicalray) [inline ap/attendance/views.py]: not all(x in filtered for x in qs)- Only deletes if the queryset has been filtered.

File: ap/leaveslips/models.py [hunk 6]
```diff
@@ -88,43 +87,60 @@ def create(self, force_insert=False, force_update=False):
...
     def get_update_url(self):
         return reverse('leaveslips:individual-update', kwargs={'pk': self.id})
 
-    def _late(self):
-        end_date = self.events.all().order_by('-end')[0].end
-        if self.submitted > end_date+timedelta(days=2):
-            return True
-        else:
-            return False
+    # def _late(self):
```
Comment (radicalray) [inline ap/leaveslips/models.py]: Check late

File: ap/leaveslips/models.py [hunk 7]
```diff
@@ -88,43 +87,60 @@ def create(self, force_insert=False, force_update=False):
...
-    @property
-    def get_start(self):  # determines the very first date of all the events
-        events=self.events.all()
-        start=datetime.now()
-        for event in events:
-            if event.start < start:
-                start=event.start
-        return start
-
+    # @property
```
Comment (radicalray) [inline ap/leaveslips/models.py]: Delete

File: ap/meal_seating/autofixtures.py [hunk 8]
```diff
@@ -1,91 +1,91 @@
-import random
+# import random
```
Comment (radicalray) [inline ap/meal_seating/autofixtures.py]: Uncomment or delete

File: ap/schedules/forms.py [hunk 9]
```diff
@@ -15,39 +16,33 @@ class EventForm(forms.ModelForm):
...
     class Meta:
         model = Event
-        fields = ('type', 'name', 'code', 'description', 'classs', 'monitor', 'term', 'start', 'end')
+        fields = ('type', 'name', 'code', 'description', 'class_type', 'monitor', 'start', 'end', 'day', 'weekday')
         widgets = { 'start': DateTimePicker(options={'format': 'MM/DD/YYYY HH:mm'}),
                     'end': DateTimePicker(options={'format': 'MM/DD/YYYY HH:mm'}) }
 
 
-class EventGroupForm(forms.ModelForm):
+# class WeeklyEventsForm(forms.ModelForm):
```
Comment (radicalray) [inline ap/schedules/forms.py]: Delete

File: ap/schedules/models.py [hunk 10]
```diff
@@ -56,125 +65,211 @@ class Event(models.Model):
...
 
-    def __unicode__(self):
-        return "[%s] %s" % (self.start.strftime('%m/%d'), self.name)
+    # gets the week from an absolute date of the current term.
+    def week_from_date(self, date):
+        return Term.current_term().term_week_of_date(date)
 
 
-class EventGroup(models.Model):
+    # def _week(self):
```
Comment (radicalray) [inline ap/schedules/models.py]: Delete

File: ap/schedules/urls.py [hunk 11]
```diff
@@ -11,6 +11,6 @@
     url(r'event/(?P<pk>\d+)/update/$', views.EventUpdate.as_view(), name='event-update'),
     url(r'event/(?P<pk>\d+)/delete/$', views.EventDelete.as_view(), name='event-delete'),
     url(r'event/(?P<term>(Fa|Sp)\d{2})/$', views.TermEvents.as_view(), name='term-events'),
-    url(r'eventgroup/create/$', views.EventGroupCreate.as_view(), name='eventgroup-create'),
-    url(r'eventgroup/(?P<pk>\d+)/$', views.EventGroupDetail.as_view(), name='eventgroup-detail'),
+    # url(r'weeklyevents/create/$', views.WeeklyEventsCreate.as_view(), name='weeklyevents-create'),
```
Comment (radicalray) [inline ap/schedules/urls.py]: Delete

File: ap/schedules/views.py [hunk 12]
```diff
@@ -37,58 +37,58 @@ def get_queryset(self):
         return Schedule.objects.filter(trainee=trainee).filter(term=Term.current_term())
 
 
-class EventGroupCreate(generic.FormView):
-    template_name = 'schedules/eventgroup_create.html'
-    form_class = EventGroupForm
+# class WeeklyEventsCreate(generic.FormView):
```
Comment (radicalray) [inline ap/schedules/views.py]: Delete

File: ap/services/admin.py [hunk 13]
```diff
@@ -1,7 +1,52 @@
-from django.contrib import admin
+# from django.contrib import admin
```
Comment (radicalray) [inline ap/services/admin.py]: Delete??

File: ap/services/fixtures/initial_data.json [hunk 14]
```diff
@@ -0,0 +1,2855 @@
+[{
```
Comment (radicalray) [inline ap/services/fixtures/initial_data.json]: Delete

File: ap/terms/models.py [hunk 15]
```diff
@@ -78,12 +79,32 @@ def set_current_term(term):
...
+ 
     @staticmethod
     def decode(code):
         """ Decode term shorthand (e.g. Sp15) """
         return Term.objects.filter(year__endswith=code[2:]).get(season__startswith=code[:2])
 
+    def is_date_within_term(self, date):
+        return date >= self.start and date <= self.end
+
+    def startdate_of_week(self, week):
```
Comment (radicalray) [inline ap/terms/models.py]: Reconsider names

File: ap/accounts/autofixtures.py [hunk 16]
```diff
@@ -1,77 +0,0 @@
-import random
```
Comment (radicalray) [inline ap/accounts/autofixtures.py]: Bring this back.

File: ap/ap/settings/dev.py [hunk 17]
```diff
@@ -4,7 +4,7 @@
 TEMPLATE_DEBUG = True
 
 INSTALLED_APPS += ('autofixture',
-                   'debug_toolbar',
+                   # 'debug_toolbar',
```
Comment (radicalray) [inline ap/ap/settings/dev.py]: Why is this commented out? Don't we still need to debug?

File: ap/ap/settings/local.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
+from .base import *
```
Comment (radicalray) [inline ap/ap/settings/local.py]: whoa, delete this file. We have a dev.py that does this.

File: ap/ap/urls.py [hunk 19]
```diff
@@ -87,6 +87,7 @@
     url(r'^api/', include(router.urls)),
     url(r'^api/', include(attendance_router.urls)),
     #third party
+    url(r'^docs/', include('rest_framework_swagger.urls')),
```
Comment (radicalray) [inline ap/ap/urls.py]: Does this actually work?! O.O

File: ap/attendance/templates/attendance/attendance_detail.html [hunk 20]
```diff
@@ -0,0 +1,429 @@
+{% extends "base.html" %}
```
Comment (radicalray) [inline ap/attendance/templates/attendance/attendance_detail.html]: Delete this file. According to @davidhtien, no longer needed.

File: ap/attendance/templates/attendance/attendance_react.html [hunk 21]
```diff
@@ -11,18 +11,31 @@
...
-	var Slips = JSON.parse('{{leaveslips_bb}}'.replace(/&#39;/g, "'").replace(/&quot;/g,'"'));
-	var Trainee = JSON.parse('{{trainee_bb}}'.replace(/&#39;/g, "'").replace(/&quot;/g,'"'));
-</script>
-<!-- figure out frontend dependencies in bower -->
-<script type="text/javascript" src="{% static "js/date_fns.min.js" %}"></script>
-<script type="text/javascript" src="{% static "js/underscore-min.js" %}"></script>
-<script type="text/javascript" src="{% static "js/index.js" %}"></script>
+	<script>
+	    var Events = [
+		    {% for event in trainee.events %}
```
Comment (radicalray) [inline ap/attendance/templates/attendance/attendance_react.html]: I wonder if we can write a template filter in django to just serialize it directly into a json object so we don't have to print this out. This will print out like this:

var Events = [
{...},
{...},
];

The last comma will break in some browsers.

File: ap/attendance/views.py (hunk 5 above)
Comment (radicalray) [inline ap/attendance/views.py]: @schai6 Did you add the comment?

File: ap/attendance/views.py [hunk 22]
```diff
@@ -89,49 +79,13 @@ class AllRollViewSet(BulkModelViewSet):
     serializer_class = RollSerializer
     filter_backends = (filters.DjangoFilterBackend,)
     filter_class = RollFilter
-    def get_queryset(self):
```
Comment (radicalray) [inline ap/attendance/views.py]: How come all this is deleted? Just to make the broken attendance work for now? Maybe you can just comment it out iwth a todo.

File: ap/leaveslips/admin.py [hunk 23]
```diff
@@ -54,10 +54,10 @@ def make_denied(modeladmin, request, queryset):
 class IndividualSlipAdmin(admin.ModelAdmin):
     fieldsets = (
         (None, {
-            'fields': ('trainee',('type', 'status',), 'description', 'comments', ('texted', 'informed',),'events','TA',)
+            'fields': ('trainee',('type', 'status',), 'description', 'comments', ('texted', 'informed',),'rolls','TA',)
```
Comment (radicalray) [inline ap/leaveslips/admin.py]: events should still work b/c we have the 
@property
 def events(self):

Line 124 in models.py in leaveslips

File: ap/leaveslips/forms.py [hunk 24]
```diff
@@ -7,7 +7,7 @@
 class IndividualSlipForm(forms.ModelForm):
     class Meta:
         model = IndividualSlip
-        fields = ['type', 'description', 'comments', 'texted', 'informed', 'events']
+        fields = ['type', 'description', 'comments', 'texted', 'informed', 'rolls']
```
Comment (radicalray) [inline ap/leaveslips/forms.py]: Can we add a TODO to support events here in the future as well? You should be able to submit a IndividualSlipForm with events rather than rolls.

File: ap/leaveslips/models.py [hunk 25]
```diff
@@ -88,44 +87,52 @@ def create(self, force_insert=False, force_update=False):
         super(LeaveSlip, self).save(force_insert, force_update)
         self.old_status = self.status
 
+    # deletes dummy roll under leave slip.
+    @receiver(pre_delete)
+    def delete_individualslip(sender, instance, **kwargs):
```
Comment (radicalray) [inline ap/leaveslips/models.py]: If this only affects individual leaveslip, should it be in the IndividualSlip class?

File: ap/leaveslips/models.py [hunk 26]
```diff
@@ -88,44 +87,52 @@ def create(self, force_insert=False, force_update=False):
...
+    rolls = models.ManyToManyField(Roll, related_name='leaveslips')
 
     def get_update_url(self):
         return reverse('leaveslips:individual-update', kwargs={'pk': self.id})
 
-    def _late(self):
-        end_date = self.events.all().order_by('-end')[0].end
-        if self.submitted > end_date+timedelta(days=2):
+    @property
+    def late(self):
```
Comment (radicalray) [inline ap/leaveslips/models.py]: Will we need this for Group Leaveslip? If so, move it to Leaveslip class instead of just keeping it in IndividualLeaveslip

File: ap/leaveslips/models.py [hunk 27]
```diff
@@ -149,7 +156,7 @@ def _events(self):
 class IndividualSlipForm(forms.ModelForm):
```
Comment (radicalray) [inline ap/leaveslips/models.py]: Why is this form defined twice?

ap/leaveslips/forms.py also has it.

Remove one instance so we don't have code duplication.

File: ap/manage.py [hunk 28]
```diff
@@ -3,7 +3,7 @@
 import sys
 
 if __name__ == "__main__":
-    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "ap.settings.base")
+    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "ap.settings.local")
```
Comment (radicalray) [inline ap/manage.py]: change this back to base

File: ap/schedules/forms.py [hunk 29]
```diff
@@ -15,39 +16,25 @@ class EventForm(forms.ModelForm):
 
     class Meta:
         model = Event
-        fields = ('type', 'name', 'code', 'description', 'classs', 'monitor', 'term', 'start', 'end')
+        fields = ('type', 'name', 'code', 'description', 'class_type', 'monitor', 'start', 'end', 'day', 'weekday')
```
Comment (radicalray) [inline ap/schedules/forms.py]: should not have class_type. This is an event form which is not a class.

File: ap/schedules/forms.py [hunk 30]
```diff
@@ -15,39 +16,25 @@ class EventForm(forms.ModelForm):
...
-    repeat = forms.MultipleChoiceField(choices=DAYS, help_text="Which days this event repeats on")
+    repeat = forms.MultipleChoiceField(choices=WEEKDAYS, help_text="Which days this event repeats on")
     duration = forms.IntegerField(help_text="How many weeks this event repeats for")
     active_trainees = Trainee.objects.filter(is_active=True)
     trainees = ModelSelect2MultipleField(queryset=active_trainees, required=False, search_fields=['^first_name', '^last_name'])
 
     class Meta:
         model = Event
-        fields = ('type', 'name', 'code', 'description', 'classs', 'monitor', 'term', 'start', 'end')
+        fields = ('type', 'name', 'code', 'description', 'class_type', 'monitor', 'start', 'end', 'day', 'weekday')
```
Comment (radicalray) [inline ap/schedules/forms.py]: Same here

File: ap/schedules/forms.py [hunk 31]
```diff
@@ -15,39 +16,25 @@ class EventForm(forms.ModelForm):
...
-        ('1', 'Tuesday'),
-        ('2', 'Wednesday'),
-        ('3', 'Thursday'),
-        ('4', 'Friday'),
-        ('5', 'Saturday'),
-        ('6', "Lord's Day"),
-    )
-
-    repeat = forms.MultipleChoiceField(choices=DAYS, help_text="Which days this event repeats on")
+    repeat = forms.MultipleChoiceField(choices=WEEKDAYS, help_text="Which days this event repeats on")
```
Comment (radicalray) [inline ap/schedules/forms.py]: We should have an option for them to indicate if event is one off or weekly repeating. And if weekly repeating, then which weekday it repeats on. repeat is too ambiguous of a name. Call it weekday.

File: ap/schedules/models.py [hunk 32]
```diff
@@ -56,125 +65,212 @@ class Event(models.Model):
...
+        end_week = weeks[len(weeks)-1]
+        return Term.current_term().start + timedelta(weeks=end_week - 1)
 
     def todays_events(self):
         today = datetime.combine(date.today(), time(0,0))
         tomorrow = today + timedelta(days=1)
         return self.events.filter(start__gte=today).filter(end__lte=tomorrow).order_by('start')
 
-    class Meta:
+    # class Meta:
```
Comment (radicalray) [inline ap/schedules/models.py]: Why is this removed? A trainee should still have one one schedule per term. So you can't add same schedule to user multiple times.

File: ap/schedules/models.py [hunk 33]
```diff
@@ -56,125 +65,212 @@ class Event(models.Model):
...
+"""
+    objects = models.Manager() # the default manager
+
+    # attach our custom managers:
+    current_term = models.CurrentTermManager()
+
+
+
+# For whole training calculation
+class CurrentTermManager(models.Manager):
```
Comment (radicalray) [inline ap/schedules/models.py]: Is this still used? If not, delete.

File: ap/schedules/serializers.py [hunk 34]
```diff
@@ -4,34 +4,49 @@
...
+    fields = '__all__'
 
 class ScheduleFilter(filters.FilterSet):
-    class Meta:
-        model = Schedule
-        fields = ['id','trainee','term','events']
\ No newline at end of file
+  class Meta:
+    model = Schedule
+    fields = ['id','trainees','weeks']
```
Comment (radicalray) [inline ap/schedules/serializers.py]: should still have events.

File: ap/attendance/views.py (hunk 22 above)
Comment (schai6) [inline ap/attendance/views.py]: It was an attempt to write OR for filters. Unnecessary.

File: ap/ap/urls.py (hunk 19 above)
Comment (schai6) [inline ap/ap/urls.py]: Yeah it works. If you go to localhost:8000/docs

File: ap/leaveslips/admin.py (hunk 23 above)
Comment (schai6) [inline ap/leaveslips/admin.py]: Hmmm it doesn't work with events.

File: ap/badges/views.py [hunk 35]
```diff
@@ -60,6 +57,7 @@ def batch(request):
             b.middlename = middle
             b.lastname = last
             b.term = Term.current_term()
+            b.term_created = Term.current_term()
```
Comment (radicalray) [inline ap/badges/views.py]: remove

File: ap/schedules/models.py [hunk 36]
```diff
@@ -240,37 +240,6 @@ def __unicode__(self):
...
-
-
-# def get(self):
-    # check priorities on each schedule and collate them into 1 schedule with all the events
-    # make an artificial schedule (composite of all the schedules)
-
-    # return artificial schedule
+    @staticmethod
+    def current_term_schedules():
+        return Schedule.objects.filter(season=Term.current_term().season, is_deleted=False)
```
Comment (radicalray) [inline ap/schedules/models.py]: check is active as well?

File: ap/badges/models.py [hunk 37]
```diff
@@ -46,7 +46,7 @@ class Badge(models.Model):
     # for defining images' paths after dropping it through dropzonejs
     def get_upload_path(self, filename):
         path = "badges/"
-        
+        print 'self', self
```
Comment (radicalray) [inline ap/badges/models.py]: delete this as well

**Comment by radicalray** [comment]
Good job @schai6 ! Merging!
